kore/src/Kore/Internal/SideCondition.hs

```diff
@@ -200,80 +202,70 @@ instance From (SideCondition variable) (MultiAnd (Predicate variable)) where
-    From (MultiAnd (Predicate variable)) (SideCondition variable)
-    where
-    from multiAnd =
-        let (assumedTrue, Assumptions termReplacements predicateReplacements) =
-                simplifyConjunctionByAssumption multiAnd
-                    & extract
-            predicateReplacementsAsTerms =
-                mapKeysAndValues
-                    Predicate.fromPredicate_
-                    predicateReplacements
-            replacements =
-                termReplacements
-                    <> predicateReplacementsAsTerms
-         in SideCondition
-                { assumedTrue
-                , replacements
-                , definedTerms = mempty
-                }
-    {-# INLINE from #-}
-
 instance
     InternalVariable variable =>
     From (SideCondition variable) (Predicate variable)
     where
     from = toPredicate
     {-# INLINE from #-}
 
-instance
-    InternalVariable variable =>
-    From (Predicate variable) (SideCondition variable)
-    where
-    from = fromPredicate
-    {-# INLINE from #-}
-
-instance
-    InternalVariable variable =>
-    From (Condition variable) (SideCondition variable)
-    where
-    from = fromCondition
-    {-# INLINE from #-}
-
 instance
     InternalVariable variable =>
     From (SideCondition variable) (Condition variable)
     where
     from = Condition.fromPredicate . toPredicate
     {-# INLINE from #-}
 
-top :: InternalVariable variable => SideCondition variable
-top =
+{- | Smart constructor for creating a 'SideCondition' by just assuming
+ a conjunction of predicates to be true.
+-}
+assumeTrue ::
+    MultiAnd (Predicate variable) ->
+    SideCondition variable
+assumeTrue assumedTrue =
     SideCondition
-        { assumedTrue = MultiAnd.top
-        , definedTerms = mempty
-        , replacements = mempty
+        { assumedTrue
+        , replacements = HashMap.empty
+        , definedTerms = HashSet.empty
         }
 
--- TODO(ana.pantilie): Should we look into removing this?
+{- | Assumes a single 'Predicate' to be true in the context of another
+ 'SideCondition'.
+ Does not modify the replacement table or the set of defined terms.
+-}
+addPredicate ::
+    Ord variable =>
+    Predicate variable ->
+    SideCondition variable ->
+    SideCondition variable
+addPredicate predicate =
+    addPredicates (MultiAnd.singleton predicate)
 
--- | A 'top' 'Condition' for refactoring which should eventually be removed.
-topTODO :: InternalVariable variable => SideCondition variable
-topTODO = top
+{- | Assumes a conjunction of 'Predicate's to be true in the context
+ of another 'SideCondition'.
+ Does not modify the replacement table or the set of defined terms.
+-}
+addPredicates ::
+    Ord variable =>
+    MultiAnd (Predicate variable) ->
+    SideCondition variable ->
+    SideCondition variable
+addPredicates predicates sideCondition =
+    sideCondition
+        { assumedTrue =
+            predicates <> assumedTrue sideCondition
+        }
```

Comment:
I would have expected to see `assumeTrue predicates <> sideCondition` here.

Reply:
`SideCondition` doesn't have a `Semigroup` instance and I kept it that way because I think adding one now would be pretty confusing, but maybe I'm wrong. I think people would assume the instance combined every field of the `SideCondition`, when it would just combine the `assumedTrue` (maybe also the `definedTerms`?) but just keep _one_ of the `replacements`.

Reply:
I see, that reason makes sense, for not having the `Semigroup` instance. Could we add a function like this instead of exposing the fields of `SideCondition`?

``` hs
addAssumptions :: MultiAnd (Predicate variable) -> SideCondition variable -> SideCondition variable
```

I think, now more than ever, it might not be a good idea to expose the fields directly.